webviz_config/_oauth2.py

```diff
@@ -0,0 +1,171 @@
+        """Check access token existence in session cookie before every request.
+        If it does not exist, the browser is redirected to /login endpoint.
+
+        If access token exists, its expiration date is checked in session cookie.
+        If the current date exceeds its expiration date, a new access token is
+        retrieved and set in the session cookie. A new expiration date is also
+        set in the session cookie.
+        """
+
+        # pylint: disable=inconsistent-return-statements
+        @self._app.before_request
+        def _check_access_token():  # type: ignore[no-untyped-def]
+            # The session of the request does not contain access token, redirect to /login
+            is_redirected, redirect_url = self.is_empty_token()
+            if is_redirected:
+                return flask.redirect(redirect_url)
+
+            # The session contains access token, check its expiration date
+            self.check_and_set_token_expiry()
+
+    @staticmethod
+    def is_empty_token() -> Tuple[bool, str]:
+        if (
+            not flask.session.get("access_token")
+            and flask.request.path != "/login"
+            and flask.request.path != "/auth-return"
+        ):
+            login_uri = replace_http_with_https(get_login_uri(flask.request.url_root))
+            return True, login_uri
+
+        return False, ""
+
+    def check_and_set_token_expiry(self) -> None:
+        if flask.session.get("access_token"):
+            expiration_date = flask.session.get("expiration_date")
+            current_date = datetime.datetime.now()
+            if current_date > expiration_date:
+                # Access token has expired
+                print("Access token has expired.")
+                if not self._accounts:
+                    self._accounts = self._msal_app.get_accounts()
+                renewed_tokens_result = self._msal_app.acquire_token_silent(
+                    scopes=[self._scope], account=self._accounts[0]
+                )
+                expires_in = renewed_tokens_result.get("expires_in")
+                new_expiration_date = datetime.datetime.now() + datetime.timedelta(
+                    seconds=expires_in - 60
+                )
+                print("New access token expiration date:", new_expiration_date)
+
+                # Set new expiration date in the session
+                flask.session["expiration_date"] = new_expiration_date
+
+                # Set new access token in the session
+                flask.session["access_token"] = renewed_tokens_result.get(
+                    "access_token"
+                )
+
+
+def get_login_uri(url_root: str) -> str:
+    return replace_http_with_https(url_root + "login")
+
+
+def get_auth_redirect_uri(url_root: str) -> str:
+    return replace_http_with_https(url_root + "auth-return")
+
+
+def replace_http_with_https(url: str) -> str:
+    return url.replace("http://", "https://")
```

Comment:
Webviz on `localhost` will probably soon go from `https` to `http`. See [webviz-config/issues/332](https://github.com/equinor/webviz-config/issues/332) and [webviz-ert/issues/87](https://github.com/equinor/webviz-ert/issues/87).

I guess this might create an issue with this solution when running on `localhost` in the future, or?

Reply:
The redirect URI for Azure AD app may be fine. But some logics should be changed (small change). We can wait until `localhost` is moved from `https` to `http` if you think this is the best route.